File: GbaGameEngine/src/game/scripts/MidiFallOfFall.cpp
#include "MidiFallOfFall.h"
#include "engine/gba/audio/GBADmgSound.h"
#include "engine/gba/audio/GBADmgMidiPlayer.h"

GBA::DMG::SoundChannel2 HeldNotePrefab();
GBA::DMG::SoundChannel2 OffNotePrefab();
GBA::DMG::SoundChannel2 MakeChannelProperties(GBA::DMG::SoundChannel2 prefab, Music::Note note, int octave);
GBA::DMG::Midi::NoteEvent MakeNoteEvent(u32 deltaTick, GBA::DMG::SoundChannel2 properties);

GBA::DMG::SoundChannel2 HeldNote = HeldNotePrefab();
GBA::DMG::SoundChannel2 OffNote = OffNotePrefab();

// BPM definitions
const int TickStep8 = 15;
const int TickStep4 = 30;
const int TickStep2 = 60;

const GBA::DMG::Midi::NoteEvent NoteEvents[] = {
	MakeNoteEvent(TickStep8, MakeChannelProperties(HeldNote, Music::Note::G,	5)),
	MakeNoteEvent(TickStep8, MakeChannelProperties(HeldNote, Music::Note::A,	5)),
	MakeNoteEvent(TickStep8, MakeChannelProperties(HeldNote, Music::Note::B,	5)),
	MakeNoteEvent(TickStep8, MakeChannelProperties(HeldNote, Music::Note::B,	5)),
	MakeNoteEvent(TickStep4, MakeChannelProperties(HeldNote, Music::Note::C,	6)),
	MakeNoteEvent(TickStep4, MakeChannelProperties(HeldNote, Music::Note::D,	6)),
	MakeNoteEvent(TickStep4, MakeChannelProperties(HeldNote, Music::Note::C,	6)),
	MakeNoteEvent(TickStep4, MakeChannelProperties(HeldNote, Music::Note::B,	5)),
	MakeNoteEvent(TickStep8, MakeChannelProperties(HeldNote, Music::Note::B,	5)),
	MakeNoteEvent(TickStep2, MakeChannelProperties(HeldNote, Music::Note::G,	5)),
	MakeNoteEvent(TickStep8, MakeChannelProperties(HeldNote, Music::Note::A,	5)),
	MakeNoteEvent(TickStep8, MakeChannelProperties(HeldNote, Music::Note::B,	5)),
	MakeNoteEvent(TickStep8, MakeChannelProperties(HeldNote, Music::Note::B,	5)),
	MakeNoteEvent(TickStep4, MakeChannelProperties(HeldNote, Music::Note::G,	6)),
	MakeNoteEvent(TickStep4, MakeChannelProperties(HeldNote, Music::Note::Fs,	6)),
	MakeNoteEvent(TickStep4, MakeChannelProperties(HeldNote, Music::Note::B,	5)),
	MakeNoteEvent(TickStep2, OffNote),
};

/***********************************************************************************************/

const GBA::DMG::Midi::NoteEvent* Midi::FallOfFall::begin()
{
	return &NoteEvents[0];
}

const GBA::DMG::Midi::NoteEvent* Midi::FallOfFall::end()
{
	constexpr int TotalEvents = sizeof(NoteEvents) / sizeof(*NoteEvents);
	return begin() + TotalEvents;
}


GBA::DMG::SoundChannel2 MakeChannelProperties(GBA::DMG::SoundChannel2 prefab, Music::Note note, int octave)
{
	prefab.frequency.soundRate = GBA::DMG::NoteToRate(note, octave);
	return prefab;
};

GBA::DMG::Midi::NoteEvent MakeNoteEvent(u32 deltaTick, GBA::DMG::SoundChannel2 properties)
{
	GBA::DMG::Midi::NoteEvent note;
	note.deltaTick = deltaTick;
	note.SetChannelProperties(properties);
	return note;
};

GBA::DMG::SoundChannel2 HeldNotePrefab()
{
	GBA::DMG::SoundChannel2 heldNote;
	heldNote.control.envelopeStepTime = 3;
	heldNote.control.envelopeDirection = GBA::DMG::SquareSound::EnvelopeStepDirection::Decrease;
	heldNote.control.envelopeInitialVolume = 15;

	heldNote.frequency.soundRate = GBA::DMG::NoteToRate(Music::Note::G, 6);// FrequencyToRate(512);
	heldNote.frequency.sustain = GBA::DMG::SquareSound::Frequency::Sustain::Hold;
	heldNote.frequency.reset = true;

	return heldNote;
}

GBA::DMG::SoundChannel2 OffNotePrefab()
{
	GBA::DMG::SoundChannel2 noteOff;
	noteOff.control.envelopeStepTime = 3;
	noteOff.control.envelopeDirection = GBA::DMG::SquareSound::EnvelopeStepDirection::Decrease;
	noteOff.control.envelopeInitialVolume = 0;

	noteOff.frequency.sustain = GBA::DMG::SquareSound::Frequency::Sustain::Hold;
	noteOff.frequency.reset = true;

	return noteOff;
}

File: GbaConversionTools/DmgMidis/MidiFallOfFall.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using GbaConversionTools.Tools;

namespace GbaConversionTools.DmgMidis
{
    internal class MidiFallOfFall
    {
        readonly static DmgMidiConverter.EventChannel2 baseNoteEventPrefab = new DmgMidiConverter.EventChannel2()
        {
            control = new DmgMidiConverter.SquareSoundRegister()
            {
                envelopeDirection = DmgMidiConverter.SquareSoundRegister.EnvelopeStepDirection.Decrease,
                envelopeInitialVolume = 15,
                envelopeStepTime = 3,
                soundLength = 0,
                waveDutyCycle = DmgMidiConverter.SquareSoundRegister.WaveDutyCycle.Cycle_3_4
            },
            frequency = new DmgMidiConverter.FrequencyRegister()
            {
                soundRate = DmgMidiConverter.NoteToRate(DmgMidiConverter.Note.G, 5),
                sustain = DmgMidiConverter.FrequencyRegister.Sustain.Hold,
                reset = DmgMidiConverter.FrequencyRegister.Reset.Enabled
            }
        };

        readonly static DmgMidiConverter.EventChannel2 offNoteEventPrefab = new DmgMidiConverter.EventChannel2()
        {
            control = new DmgMidiConverter.SquareSoundRegister()
            {
                envelopeDirection = DmgMidiConverter.SquareSoundRegister.EnvelopeStepDirection.Decrease,
                envelopeInitialVolume = 0,
                envelopeStepTime = 7,
                soundLength = 0,
                waveDutyCycle = DmgMidiConverter.SquareSoundRegister.WaveDutyCycle.Cycle_3_4
            },
            frequency = new DmgMidiConverter.FrequencyRegister()
            {
                soundRate = DmgMidiConverter.NoteToRate(DmgMidiConverter.Note.G, 5),
                sustain = DmgMidiConverter.FrequencyRegister.Sustain.Hold,
                reset = DmgMidiConverter.FrequencyRegister.Reset.Enabled
            }
        };

        static DmgMidiConverter.Event MakeBasicNoteEvent(ushort deltaTick, DmgMidiConverter.EventChannel2 prefab, DmgMidiConverter.Note note, int octave)
        {
            prefab.frequency.soundRate = DmgMidiConverter.NoteToRate(note, octave);
            return new DmgMidiConverter.Event()
            {
                m_deltaTick = deltaTick,
                m_properties = prefab
            };
        }

        public static void MakeMidi(string inputPath, string outputPath)
        {
            DmgMidiConverter converter = new DmgMidiConverter();
            List<DmgMidiConverter.Event> midiEvents = new List<DmgMidiConverter.Event>();

            const int TickStep8 = 15;
            const int TickStep4 = 30;
            const int TickStep2 = 60;

            midiEvents.Add(MakeBasicNoteEvent(TickStep8, baseNoteEventPrefab, DmgMidiConverter.Note.G, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep8, baseNoteEventPrefab, DmgMidiConverter.Note.A, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep8, baseNoteEventPrefab, DmgMidiConverter.Note.B, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep8, baseNoteEventPrefab, DmgMidiConverter.Note.B, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep4, baseNoteEventPrefab, DmgMidiConverter.Note.C, 6));
            midiEvents.Add(MakeBasicNoteEvent(TickStep4, baseNoteEventPrefab, DmgMidiConverter.Note.D, 6));
            midiEvents.Add(MakeBasicNoteEvent(TickStep4, baseNoteEventPrefab, DmgMidiConverter.Note.C, 6));
            midiEvents.Add(MakeBasicNoteEvent(TickStep4, baseNoteEventPrefab, DmgMidiConverter.Note.B, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep8, baseNoteEventPrefab, DmgMidiConverter.Note.B, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep2, baseNoteEventPrefab, DmgMidiConverter.Note.G, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep8, baseNoteEventPrefab, DmgMidiConverter.Note.A, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep8, baseNoteEventPrefab, DmgMidiConverter.Note.B, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep8, baseNoteEventPrefab, DmgMidiConverter.Note.B, 5));
            midiEvents.Add(MakeBasicNoteEvent(TickStep4, baseNoteEventPrefab, DmgMidiConverter.Note.G, 6));
            midiEvents.Add(MakeBasicNoteEvent(TickStep4, baseNoteEventPrefab, DmgMidiConverter.Note.Fs, 6));
            midiEvents.Add(MakeBasicNoteEvent(TickStep4, baseNoteEventPrefab, DmgMidiConverter.Note.B, 5));
            midiEvents.Add(new DmgMidiConverter.Event() { m_deltaTick = TickStep2, m_properties = offNoteEventPrefab });

            converter.Convert(inputPath, outputPath, midiEvents);
        }
    }
}
